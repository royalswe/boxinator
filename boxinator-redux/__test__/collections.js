const boxCollections = [
    {
      "id": 27,
      "created_at": "2018-03-11 17:43:12",
      "updated_at": "2018-03-11 17:43:12",
      "name": "Roy",
      "color": "rgb(255, 0, 0)",
      "country": "Sweden",
      "weight": "45.00"
    },
    {
      "id": 38,
      "created_at": "2018-03-12 19:58:12",
      "updated_at": "2018-03-12 19:58:12",
      "name": "Bradley",
      "color": "rgb(78,40,24)",
      "country": "China",
      "weight": "8.00"
    }
];

export {boxCollections};