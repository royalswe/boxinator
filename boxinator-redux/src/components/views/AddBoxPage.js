/**
 * TODO:: backend validation should render it's error messages and not a default error.
 * find out what is considered blue color.
 * exports AddBoxPage class only for testing purpose.
 */
import React from 'react';
import classnames from 'classnames';
import reactCSS from 'reactcss'
import { connect } from 'react-redux';
import { createItem } from '../../actions/actions';
import { ChromePicker } from 'react-color';

export class AddBoxPage extends React.Component {

    state = {
        name: '',
        color: {r:241, g:112, b:0}, // default color
        weight: '',
        country: 'Sweden',
        displayColorPicker: false,
        displayMessage: false,
        errors: {}
    };

    handleChange = (e) => {
        /**
         * If there is any validation errors,
         * validation errors removes when it is valid
         */
        if (!!this.state.errors[e.target.name]) {
          let errors = Object.assign({}, this.state.errors);      
          delete errors[e.target.name];
          this.setState({
            [e.target.name]: e.target.value,
            errors
          });
        } else {
          this.setState({ [e.target.name]: e.target.value });
        }
      }

    handleColorClick = () => {
        this.setState({ displayColorPicker: !this.state.displayColorPicker })
    };

    handleColorClose = () => {
        this.setState({ displayColorPicker: false })
    };

    handleColorChange = (color) => {
        /**
         * Check if color is blue
         */
        if(color.hsv.h > 165 && color.hsv.h < 270){
            let errors = {color: "Please don't pick a blue color"};
            this.setState({errors});
        }
        else{
            // remove errors if any and set color state
            delete this.state.errors.color;
            this.setState({ color: color.rgb })
        }
    };

    handleSubmit = (e) => {
        e.preventDefault();

        /**
         * custom error messages
         */
        let errors = {};
        if(this.state.name === '') errors.receiver = "Name can't be empty";
        if(this.state.country === '') errors.color = "Country can't be empty";
        if(this.state.weight === '') errors.weight = "Weight can't be empty";
        if(this.state.weight < 0) {
            errors.weight = "Weight can't be a negative number";
            this.setState({weight: 0});
        }
        
         /**
         * Validate form and prevent submit if is not valid
         */
        this.setState({errors});
        const isValid = Object.keys(errors).length === 0;
        if(isValid){ // if form dont have errors
            const color = `rgb(${this.state.color.r}, ${this.state.color.g}, ${this.state.color.b})`;
            const { name, weight, country } = this.state;

            this.props.createItem({name, color, weight, country}).then(response => {
                // if box created successfully then reset the state
                this.setState({
                    name: '',
                    weight: '',
                    country: 'Sweden',
                    displayMessage: <p className="box-created-successfully">Box is succesfully created!</p>,
                    errors: {}
                });
                setTimeout(() => { // show success message for 4 seconds
                    this.setState({displayMessage: false})
                  }, 4000);

              }).catch(errors => {
                console.log(errors.message);
                this.setState({displayMessage: <p className="box-create-fail">{errors.message}</p>});
              });         
        }
    }

    render() {
        // styles for color picker
        const styles = reactCSS({
            'default': {
              color: {
                width: '36px',
                height: '14px',
                borderRadius: '2px',
                background: `rgb(${ this.state.color.r }, ${ this.state.color.g }, ${ this.state.color.b })`,
              },
              swatch: {
                padding: '5px',
                background: '#fff',
                borderRadius: '1px',
                boxShadow: '0 0 0 1px rgba(0,0,0,.1)',
                display: 'inline-block',
                cursor: 'pointer',
              },
              popover: {
                position: 'absolute',
                zIndex: '2',
              },
              cover: {
                position: 'fixed',
                top: '0px',
                right: '0px',
                bottom: '0px',
                left: '0px',
              },
            },
          });

          // countries select values 
          let countryOptions = ['Sweden', 'China', 'Brazil', 'Australia'].map(country => {
            return <option key={country} value={country}>{country}</option>
          });

        return (
            <div>
                <form className="add-box" onSubmit={this.handleSubmit}>
                    <h2>Add new box</h2>
            
                    <span className="error-text">{this.state.errors.receiver}</span>
                    <div className={classnames('field', { error: !!this.state.errors.receiver})}>
                        <label htmlFor="name">Name</label>
                        <input
                            name="name"
                            value={this.state.name}
                            onChange={this.handleChange}
                            id="name"
                            required
                        />
                    </div>

                    <span className="error-text">{this.state.errors.weight}</span>
                    <div className={classnames('field', { error: !!this.state.errors.weight})}>
                        <label htmlFor="weight">Weight</label>
                        <input
                            name="weight"
                            value={this.state.weight}
                            onChange={this.handleChange}
                            id="weight"
                            type="number"
                            required
                        />
                    </div>

                    <span className="error-text" >{this.state.errors.country}</span>
                    <div className={classnames('field', { error: !!this.state.errors.country})}>
                        <label htmlFor="country">Destination Country</label>
                        <select ref="countries" name="country" value={this.state.country} onChange={this.handleChange}>
                            {countryOptions}
                        </select>
                    </div>

                    <span className="error-text">{this.state.errors.color}</span> 
                    <div className={classnames('field', { error: !!this.state.errors.color})}>
                        <div>Box Color</div>
                        <div style={ styles.swatch } onClick={ this.handleColorClick }>
                        <div style={ styles.color } />
                        </div>
                        { this.state.displayColorPicker ? <div style={ styles.popover }>
                        <div style={ styles.cover } onClick={ this.handleColorClose }/>
                        <ChromePicker disableAlpha color={ this.state.color } onChange={ this.handleColorChange }/>
                        </div> : null }
                    </div>

                    <div>
                        <button>Save</button>
                    </div>
                </form>

                {!!this.state.displayMessage && <div>{this.state.displayMessage}</div>}

            </div>
        )
    }
}

export default connect(null, { createItem })(AddBoxPage);