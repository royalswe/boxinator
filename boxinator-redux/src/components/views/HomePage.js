import React from 'react';

class HomePage extends React.Component {

    render() {
        return (
            <div>
                <h2>Welcome to Boxinator!</h2>
                <p>This is a simple application build with React and Redux</p>
            </div>
        )
    }
}

export default HomePage;