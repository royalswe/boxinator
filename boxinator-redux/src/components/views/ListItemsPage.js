import React from 'react';
import { connect } from 'react-redux';
import ItemList from '../ItemList';
import PropTypes from 'prop-types';
import { fetchItems } from '../../actions/actions';


class ListItemsPage extends React.Component {
  componentDidMount(){
    this.props.fetchItems();
  }
  render() {
    console.log(this.props.items);
    
    return (
      <div>
        <h1>Box List</h1>
        <ItemList items={this.props.items} />
      </div>
    );
  }
  
}

ListItemsPage.propTypes = {
  items: PropTypes.array.isRequired,
  fetchItems: PropTypes.array.isRequired
}

function mapStateToProps(state) {
  return {
    items: state.items
  }
}

export default connect(mapStateToProps, { fetchItems })(ListItemsPage);

