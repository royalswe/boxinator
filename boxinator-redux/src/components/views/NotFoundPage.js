import React from 'react';

class HomePage extends React.Component {

    render() {
        // Just a simple 404 page
        return (
            <div>
                <h2>404</h2> 
                <h4>Could not find the page <code>{window.location.pathname}</code></h4>
            </div>
        )
    }
}

export default HomePage;