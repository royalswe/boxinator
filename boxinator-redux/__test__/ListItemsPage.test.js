import React from 'react';
import { ListItemsPage } from '../src/components/views/ListItemsPage';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { boxCollections } from './collections'

describe('<ListItemsPage /> should render box items', () => {

  // beforeEach(()=>{  })

  it('Get notified if no boxes exists', () => {
    const wrapper = render(<ListItemsPage items={ [] }/>);
    expect(wrapper.find('p').text()).toEqual('There are no boxes in collection.');
    
  });

  it('Should render a table including itÂ´s data and content', () => {
    const wrapper = render(<ListItemsPage items={ boxCollections }/>); 
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

});