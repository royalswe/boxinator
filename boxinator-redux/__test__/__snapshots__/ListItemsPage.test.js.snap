// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItemsPage /> should render box items Should render a table including itÂ´s data and content 1`] = `
<div>
  <h1>
    Box List
  </h1>
  <div>
    <table>
      <tbody>
        <tr
          class="list-group"
        >
          <th>
            Roy
          </th>
          <th>
            45.00 kilograms
          </th>
          <th
            style="background-color:rgb(255, 0, 0)"
          />
          <th>
             SEK
          </th>
        </tr>
        <tr
          class="list-group"
        >
          <th>
            Bradley
          </th>
          <th>
            8.00 kilograms
          </th>
          <th
            style="background-color:rgb(78,40,24)"
          />
          <th>
             SEK
          </th>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <th>
            045.008.00
          </th>
          <th>
            NaN
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
